camel-core/src/main/java/org/apache/camel/util/GenericCaseInsensitiveMap.java

```diff
@@ -0,0 +1,549 @@
+package org.apache.camel.util;
+
+import java.util.*;
+
+
+/**
+ * Useful Map that does not care about the case-sensitivity of keys
+ * when the key value is a String.  Other key types can be used.
+ * String keys will be treated case insensitively, yet key case will
+ * be retained.  Non-string keys will work as they normally would.
+ * <p>
+ * The internal CaseInsensitiveString is never exposed externally
+ * from this class. When requesting the keys or entries of this map,
+ * or calling containsKey() or get() for example, use a String as you
+ * normally would.  The returned Set of keys for the keySet() and
+ * entrySet() APIs return the original Strings, not the internally
+ * wrapped CaseInsensitiveString.
+ * <p>
+ * As an added benefit, .keySet() returns a case-insenstive
+ * Set, however, again, the contents of the entries are actual Strings.
+ * Similarly, .entrySet() returns a case-insensitive entry set, such that
+ * .getKey() on the entry is case insensitive when compared, but the
+ * returned key is a String.
+ * <p>
+ * <p>
+ * <p>
+ * Customizations:
+ * - Removed support for other internal map implementation than <i>HashMap</i>
+ * - Renamed to <i>GenericCaseInsensitiveMap</i>
+ * - Added <i>Cloneable, java.io.Serializable</i> interfaces support
+ *
+ * @author John DeRegnaucourt (john@cedarsoftware.com)
+ *         <br>
+ *         Copyright (c) Cedar Software LLC
+ *         <br><br>
+ *         Licensed under the Apache License, Version 2.0 (the "License");
+ *         you may not use this file except in compliance with the License.
+ *         You may obtain a copy of the License at
+ *         <br><br>
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *         <br><br>
+ *         Unless required by applicable law or agreed to in writing, software
+ *         distributed under the License is distributed on an "AS IS" BASIS,
+ *         WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *         See the License for the specific language governing permissions and
+ *         limitations under the License.
+ */
+public class GenericCaseInsensitiveMap<K, V> implements Map<K, V>, Cloneable, java.io.Serializable {
+    private final Map<K, V> map;
+
+    public GenericCaseInsensitiveMap() {
+        map = new HashMap<>();
+    }
+
+    public GenericCaseInsensitiveMap(int initialCapacity) {
+        map = new HashMap<>(initialCapacity);
+    }
+
+    public GenericCaseInsensitiveMap(Map<K, V> m) {
+        map = copy(m, new HashMap(m.size()));
+    }
+
+    protected Map<K, V> copy(Map<K, V> source, Map dest) {
+        for (Entry<K, V> entry : source.entrySet()) {
+            K key = entry.getKey();
+            K altKey;
```

Comment:
I think we could get rid of the `altKey` variable and the corresponding else block as well.